### ISSUE TITLE
[MNG-5852] mvn shell script invokes /bin/sh but requires Bash functions

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7575]
**[Jeffrey Alexander](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhalex)** opened **[MNG-5852](https://issues.apache.org/jira/browse/MNG-5852?redirect=false)** and commented

The bin/mvn script uses the "local" command which is a shell builtin of bash and similar shells, but is not required for POSIX-compliance in sh.  When I attempt to run mvn on my Solaris system, I see the following output:

```
$ ./mvn
./mvn[200]: local: not found [No such file or directory]
./mvn[201]: local: not found [No such file or directory]
...
```

Lines 200 and 201 invoke "local" to make local variables to the function.  According to "man bash", this is a shell builtin.  However, bin/mvn is invoked as:

#!/bin/sh

On most flavors of linux, this resolves to bash or dash which probably runs in a restricted environment after checking to see that its $0 is sh. But on Solaris's /bin/sh is actually ksh93 for backwards compatibility.

Since "local" is not part of a POSIX-compliant /bin/sh, depending on it in a script that is invoked with /bin/sh is a bug.

---

**Affects:** 3.3.3

**Issue Links:**
- [MNG-6095](https://issues.apache.org/jira/browse/MNG-6095) mvn script warns on Solaris
 (_**"is duplicated by"**_)
- [MNG-5837](https://issues.apache.org/jira/browse/MNG-5837) Syntax error in bin/mvn on Solaris SPARC
 (_**"supercedes"**_)

**Backported to:** [3.5.0-alpha-1](https://github.com/apache/maven/milestone/92?closed=1)

2 votes, 8 watchers

**Comment by jira-importer** [comment]
**[Thorsten Heit](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=t_heit)** commented

I also saw this output and decided to just remove the word "local" in the shell script to make it work again. So far I haven't hit any problems.

**Comment by jira-importer** [comment]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77)** commented

I am running 'OpenBSD 5.8'. '/bin/sh' also is 'ksh' and works flawlessly. Can you please confirm the following changes to 'apache-maven/src/bin/mvn' solve your issue?

```
@@ -197,8 +197,6 @@ fi
1. traverses directory structure from process work directory to filesystem root
1. first directory with .mvn subdirectory is considered project base directory
 find_maven_basedir() {
-  local basedir
-  local wdir
   basedir="$(pwd)"
   wdir="$(pwd)"
   while [ "$wdir" != '/' ] ; do
```

**Comment by jira-importer** [comment]
**[Jeffrey Alexander](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhalex)** commented

Thank you, that does work as expected!

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

Maven 3.4.0 has been dropped. See [this thread](https://mail-archives.apache.org/mod_mbox/maven-dev/201701.mbox/%3CCA%2BnPnMx-e7kGYy3Hp87v8hLGdhp1q%3DtKLx_6QuZ4kGUqHEBGcw%40mail.gmail.com%3E) for more details.

This issue will need to be re-scheduled for a Maven release in the (hopefully near) future.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [135102022763eaa62fa1d0125adab7ff207ef2f4](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commit;h=135102022763eaa62fa1d0125adab7ff207ef2f4).
